osquery/core/init.cpp

```diff
@@ -154,50 +161,97 @@ void Initializer::initDaemon() {
   }
 }
 
-void Initializer::initWorkerWatcher(const std::string& name) {
-  // The watcher will forever monitor and spawn additional workers.
-  Watcher watcher(argc_, argv_);
-  watcher.setWorkerName(name);
+void Initializer::initWatcher() {
+  // The watcher takes a list of paths to autoload extensions from.
+  auto autoload_paths = osquery::split(FLAGS_extensions_autoload, ":");
+  for (auto& path : autoload_paths) {
+    boost::trim(path);
+    Watcher::addExtensionPath(path);
+  }
 
-  if (isWorker()) {
-    // Do not start watching/spawning if this process is a worker.
-    watcher.initWorker();
-  } else {
-    do {
-      if (!watcher.watch()) {
-        // The watcher failed, create a worker.
-        watcher.createWorker();
-        watcher.resetCounters();
-      }
-    } while (watcher.ok());
+  // Add a watcher service thread to start/watch an optional worker and set
+  // of optional extensions in the autoload paths.
+  if (Watcher::countExtensions() > 0 || !FLAGS_disable_watchdog) {
```

Comment:
Wrap the thread constructor in a conditional to avoid erroring quickly without joining on the yet-to-be-allocated thread context IF the config/logger plugins do not exist. Otherwise Initializer will wait for autoloaded extensions to start (using extensions_timeout), then exit if the target/active plugins were not broadcasted. By this time terminating without joining is practical.